## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [196]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [197]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [198]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [199]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

# gross_df  = df.loc[(df["first"] == "Erik") & (df["last"] == "Gross")]
# gross_df.head()

# unit_price_series = df["unit_price"].copy()


# print(unit_price_series)

# q1 = unit_price_series.quantile(0.25)
# median = unit_price_series.quantile(0.5)
# q3 = unit_price_series.quantile(0.75)

# iqr = q3 - q1
# lower_bound = q1 + (1.5 * iqr)
# upper_bound = q3 + (1.5 * iqr)
# print("---")
# print(f'UpperBound:, {upper_bound:,.2f}, "LowerBound:", {lower_bound:,.2f}')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [200]:
# What three item categories had the most entries?
category_series  = df.value_counts("category")
category_series.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [201]:
# For the category with the most entries, which subcategory had the most entries?
consumables_df = df.loc[df["category"] == "consumables", :]
consumables_df.head(5)
top_subcategory = consumables_df.value_counts("subcategory")
top_subcategory.head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [202]:
# Which five clients had the most entries in the data?
top_five_clients_df = df.value_counts("client_id")
top_five_clients_df.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [203]:
# Store the client ids of those top 5 clients in a list.
top_five_df = pd.DataFrame(top_five_clients_df)
top_five_df
top_five_array= top_five_df.head(5).index.tolist()
top_five_array

[33615, 66037, 46820, 38378, 24741]

In [204]:
# How many total units (the qty column) did the client with the most entries order?
total_units_df = df.loc[df["client_id"] == 33615]
total_units_df
total_units_df["qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [205]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
def calculate_subtotal(row):
    return row['unit_price'] * row["qty"]

df["line_subtotal"] = df.apply(calculate_subtotal, axis = 1 )
subtotal_df = df[["client_id", "unit_price", "qty", "line_subtotal"]]
# subtotal_df = df.loc[df["client_id"] == 33615, ["client_id", "unit_price", "qty", "line_subtotal"]]

subtotal_df


client_id  unit_price   qty  line_subtotal
0          58515     1096.80   105      115164.00
1          37609       24.95    21         523.95
2          57113       13.52    39         527.28
3          46554       36.42    29        1056.18
4          92089      195.10    20        3902.00
...          ...         ...   ...            ...
54634      29043       83.13    33        2743.29
54635      41908      206.59    47        9709.73
54636      35176       65.66   475       31188.50
54637      24485        1.48   112         165.76
54638      24007        3.01  1031        3103.31

[54639 rows x 4 columns]

In [206]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping_price(row):
    _51_pds_and_up_price = 7
    _50_and_under_price = 10
    unit_weight = row["total_weight"]
    if unit_weight > 50:
        return _51_pds_and_up_price * unit_weight
    elif unit_weight <= 50:
        return _50_and_under_price * unit_weight
    
def calculate_total_weight (row):
    return (row["unit_weight"] * row["qty"])
df["total_weight"] = df.apply(calculate_total_weight, axis =1)
df["shipping_price"] = df.apply(calculate_shipping_price, axis = 1 )


shipping_price_df = df[["unit_price", "unit_weight", "qty", "total_weight", "shipping_price"]]
shipping_price_df

unit_price  unit_weight   qty  total_weight  shipping_price
0         1096.80         7.50   105        787.50         5512.50
1           24.95         1.49    21         31.29          312.90
2           13.52         1.68    39         65.52          458.64
3           36.42         1.23    29         35.67          356.70
4          195.10        46.43    20        928.60         6500.20
...           ...          ...   ...           ...             ...
54634       83.13         2.25    33         74.25          519.75
54635      206.59        11.70    47        549.90         3849.30
54636       65.66         4.16   475       1976.00        13832.00
54637        1.48        18.04   112       2020.48        14143.36
54638        3.01         2.07  1031       2134.17        14939.19

[54639 rows x 5 columns]

In [207]:
# HELP: 1 of 3: LINE PRICE DOES NOT MATCH!!!

# Create a column for the total price using the line_subtotal and the shipping_price along with a sales tax of 9.25%
def calculate_total_price(row):
    subtotal = row["line_subtotal"]
    # sales_tax =  subtotal * .0925
    sales_tax =  1.0925

    shipping = row["shipping_price"]
    total_cost = (subtotal +  shipping) * sales_tax
    # return f'{total_cost:.2f}'
    return round(total_cost, 2)


df["line_price"] = df.apply(calculate_total_price, axis = 1)
df["line_price"] = df["line_price"].astype('float')
total_price_df = df[["line_subtotal", "shipping_price", "line_price"]]
total_price_df

line_subtotal  shipping_price  line_price
0          115164.00         5512.50   131839.08
1             523.95          312.90      914.26
2             527.28          458.64     1077.12
3            1056.18          356.70     1543.57
4            3902.00         6500.20    11364.40
...              ...             ...         ...
54634        2743.29          519.75     3564.87
54635        9709.73         3849.30    14813.24
54636       31188.50        13832.00    49184.90
54637         165.76        14143.36    15632.71
54638        3103.31        14939.19    19711.43

[54639 rows x 3 columns]

In [208]:
# Create a column for the cost of each line using unit_cost, qty, and
# shipping_price (assume the shipping cost is exactly what is charged to the client).
def calculate_line_cost(row):
    return (row["unit_cost"] * row["qty"]) + row["shipping_price"]
df["line_cost"] = df.apply(calculate_line_cost, axis = 1)
df["line_cost"] = df["line_cost"].astype('float')
# line_cost_df = df[["line_subtotal", "shipping_price", "line_price"]]
df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   
3            3        1056.18         35.67          356.70     1543.57   
4            1        3902.00        928.60         6500.20    11364.40   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [209]:
# HELP: 2 of 3: LINE_PRICE IS STILL OFF:

# Create a column for the profit of each line using line cost and line price
def calculate_profit(row):
    return row["line_price"] - row["line_cost"] 
df["profit"] = df.apply(calculate_profit, axis = 1)
df.head(5)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   
3        1056.18         35.67          356.70     1543.57    1077.35   
4        3902.00        928.60         6500.20    11364.40    8663.60   

     profit  
0  46242.03  
1    284.47  
2    311.94  
3    466.22  
4   2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [210]:
# HELP: 3 of 3:  LINE_PRICE TOTAL IS STILL OFF

# Check your work using the totals above
subgroup_df = df[["order_id","line_price"]]
subgroup_df
totals = {'Client'}
order_ids = [2742071, 2173913, 6128929]
client_id_df = subgroup_df\
    .loc[subgroup_df["order_id"].isin(order_ids), :].copy().set_index("order_id")

# Return a new datafrom
client_id_df

# Sum the line_price per order_id
order_totals = client_id_df.groupby("order_id")["line_price"].sum()

# Return a series
order_totals
# Loop through the Series
for order_id, total in order_totals.items():
    print(f"Order ID: {order_id}, Total: {total:,.2f}")


Order ID: 2173913, Total: 162,388.71
Order ID: 2742071, Total: 152,811.89
Order ID: 6128929, Total: 923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [211]:
# How much did each of the top 5 clients (top_five_array) by quantity (from Part1) spend? Check your work from Part 1 for client ids.
top_five_array
clients_df = df[df.client_id.isin(top_five_array)]
clients_df.groupby("client_id")["line_price"].sum()
# top_five_spend_df = df[df["client_id"].isin(top_five_array)].groupby("client_id")["line_price"].sum()
# top_five_spend_df

# client_subgroup_df = df[["client_id","line_price"]]
# client_subgroup_df.iloc[client_subgroup_df["client_id"]]
# def return_line_price(row):
#     for client in top_five_array:
#         print(f'{row["client_id"]}: ${row["line_price"]:,.2f}')
# top_five_spend_df = df.apply(return_line_price, axis = 1)

# results_df = df.groupby("client_id")["line_price"].sum()
# results_df
# top_five_spend_df.groupby

# top_clients = df.groupby("client_id")["qty"].sum().sort_values(ascending=False).head(5)
# top_client_ids = top_clients.index.tolist()
# top_client_ids


client_id
24741    82268892.04
33615     8377308.53
38378    12906550.87
46820     9743794.33
66037    10259514.80
Name: line_price, dtype: float64

In [212]:
clients_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'line_subtotal', 'total_weight', 'shipping_price',
       'line_price', 'line_cost', 'profit'],
      dtype='object')

In [213]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
client_total_summary_df = clients_df[['client_id', 'qty', 'shipping_price', 'line_price', 'line_cost', 'profit']].groupby("client_id").sum().reset_index()
client_total_summary_df

client_id     qty  shipping_price   line_price    line_cost       profit
0      24741  239862      5126448.37  82268892.04  45688899.71  36579992.33
1      33615   64313      1828984.89   8377308.53   6175313.91   2201994.62
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.33   7007191.64   2736602.69
4      66037   43018      1395151.85  10259514.80   7004482.98   3255031.82

In [214]:
# Format the data and rename the columns to names suitable for presentation.

# Rename the columns to reflect the change in the money format. 
updated_names_df = client_total_summary_df.rename(
    columns={"qty": "Units", 
             "shipping_price": "Shipping (millions)", 
             "line_price": "Total Revenue (millions)", 
             "line_cost": "Total Cost (millions)",
             "profit": "Total Profit (millions)"})

updated_names_df
# Define the money column names array. 

column_names = [ 
                 "Shipping (millions)", 
                 "Total Revenue (millions)", 
                 "Total Cost (millions)",
                 "Total Profit (millions)"
                 ]

# Define a function that converts a dollar amount to millions.
def scale_to_millions(value):
    return value / 1000000

# Apply the currency_format_millions function to each money column value. 
for name in column_names:
    # updated_names_df[name] = updated_names_df[name]/ 1000000
    
    # NOTE: You must store your value back in the table
    updated_names_df[name] = updated_names_df[name].apply(scale_to_millions)


updated_names_df


client_id   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
1      33615   64313             1.828985                  8.377309   
2      38378   73667             3.429455                 12.906551   
3      46820   75768             1.601449                  9.743794   
4      66037   43018             1.395152                 10.259515   

   Total Cost (millions)  Total Profit (millions)  
0              45.688900                36.579992  
1               6.175314                 2.201995  
2               9.634721                 3.271830  
3               7.007192                 2.736603  
4               7.004483                 3.255032

In [215]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = updated_names_df.sort_values("Total Profit (millions)", ascending=False)
sorted_df

client_id   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862             5.126448                 82.268892   
2      38378   73667             3.429455                 12.906551   
4      66037   43018             1.395152                 10.259515   
3      46820   75768             1.601449                  9.743794   
1      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
0              45.688900                36.579992  
2               9.634721                 3.271830  
4               7.004483                 3.255032  
3               7.007192                 2.736603  
1               6.175314                 2.201995